libtiff/libtiff_ctypes.py

```diff
@@ -563,14 +563,9 @@ def read_image(self, verbose=False):
                     raise IOError("Unexpected PlanarConfig = %d" % planar_config)
             size = arr.nbytes
 
-            if compression == COMPRESSION_NONE:
```

Comment:
Unfortunately, we don't have unit tests to test the consequences of this change, so I would be conservative here.

Do you know what will be the performance/memory usage impacts to fully switching to encoded versions of read/write functions?

Would also the following approach work (not tested):
```
fillorder = self.GetField('FillOrder')
if not (sys.byteorder == "little" and fillorder ==  FILLORDER_LSB2MSB) # system and data endianess match
   and compression == COMPRESSION_NONE:
    ....
```
that should have less impact to performance and possibly for other users reading/writing raw data.

What do you think?

Reply:
Fair enough, it's good to check that it's not causing performance regressions. I just had a look. Long story short: on the write, I didn't notice any change, but on the read, it's about 2% slower. IHMO, it's acceptable for the sake of ensuring that all the images are properly handled, while having simpler code.

To be honest, looking at the libtiff code (https://gitlab.com/libtiff/libtiff/blob/master/libtiff/tif_read.c#L512), I was pretty convinced that it could even be a tiny bit faster, because there is an explicit "shortcut" in the C code which does approximately the same checks as in Python (ie, compression is None + no byte swap). Such C code should be faster  than the equivalent Python code, so that should lead to faster code. I was wrong.

Firstly, I tried on some example TIFF images found in the libtiff test cases. However, all the image are tiny (the biggest I found is 140Kb). On such images, the difference is always negligible. So, I tried on a 5Mpx uint16 image (11Mb), and the results (on my laptop, with Python 2.7) with the original code are:
 * read: 11.9 +- 0.2 ms
 * write: 3.9 +- 0.2 ms
With the new code:
 * read: 12.1 +- 0.2ms
 * write: 3.9 +- 0.2 ms

So that's around 2% performance decrease when reading. I've dug a little further, and interestingly, half of the extra time  is due to the extra `self.` in the loop. So changing from `self.ReadEncodedStrip()` to `ReadStrip = self.ReadEncodedStrip... ReadStrip()`, saves about 0.1ms. That's because the image has 2560 rows (and there is one strip per row). This loop is  probably also what explains why the C code is slower: because the checks are done inside the loop instead of outside the loop, so they are done 2560 times, instead of once in Python.
Surprisingly (to me), when writing, there is just a single long strip. So there is no loop, and that's what explains there is no time difference.

So, we could consider putting back the "trick" to avoid the `self.` dereference inside the loop to reduce even further the performance penalty.

For info, here is my (extremely sophisticated :-P) test code:
```
from libtiff import TIFF
imfn = "test-big-16b-gs.tiff"
tif = TIFF.open(imfn, mode='r')
%timeit -n1000 image = tif.read_image(); tif.SetDirectory(0)
```
I can provide you the "big image" by email if you want.

Reply:
Thanks for timing the different approaches!
I suggest replacing the `if compression == ...` blocks with `ReadStrip = self.ReadEncodedStrip` and `WriteStrip = self.WriteEnc...`, respectively, and elsewhere use the original code. The patch will be smaller and there is tiny performance gain.